CIP-plutus-modules/README.md

```diff
@@ -0,0 +1,1599 @@
+to force delayed closures automatically where the value is required,
+thus enabling explicit `Force` operations to be omitted. This would
+effectively turn UPLC into a lazy programming language. That would
+enable this problem to be solved more easily, but  at the cost of
+reversing a rather fundamental design decision in UPLC--and probably
+making the CEK machine a little bit slower, for all programs.
+
+Thus it appears that there is no good way of using UPLC's existing
+lazy evaluation to observe use of script arguments.
+
+### Value Scripts
+
+This section discusses the 'value scripts' variation.
+
+The main specification in this CIP represents a `Script` that imports
+modules as compiled code applied to a list of `ScriptHash`es. To
+prepare such a script for running, `resolveScriptDependencies`
+replaces each hash by the term it refers to, and builds nested
+applications of the compiled code to the arguments. These applications
+must be evaluated by the CEK machine *before* the script proper begins
+to run. Moreover, each imported module is itself translated into such
+a nested application, which must be evaluated before the module is
+passed to the client script. In a large module hierarchy this might
+cause a considerable overhead before the script proper began to
+run. Worst of all, if a module is used *several times* in a module
+dependency tree, then it must be evaluated *each time* it is
+used. Resolving module dependencies traverses the entire dependency
+*tree*, which may be exponentially larger than the dependency *dag*.
+
+The value script variation addresses this problem head on. Scripts are
+converted directly into CEK-machine values that can be invoked at low
+cost. Each script is converted into a value only once, no matter how
+many times it is referred to, saving time and memory when modules
+appear several times in a module hierarchy.
+
+On the other hand it does restrict the syntactic form of
+scripts. Scripts are restricted to be syntactic lambda expressions,
+binding their script arguments at the top-level. This is not so
+onerous. But inside those λs, there must also be a syntactic
+value. For example, consider a module represented by a tuple, whose
+components represent the exports of the module. Then all of those exports
+need to be syntactic values--an exported value could not be computed
+at run-time, for example using an API exported by another
+module. While many module exports are functions, and so naturally
+written as λ-expressions (which are values), this restriction will be
+onerous at times.
+
+This method does require opening up the API of the CEK machine, so
+that CEK values can be constructed in other modules, and introducing a
+way to run the machine starting from a given configuration. So it
+requires more invasive changes to the code than the main
+specification.
+
+#### `ScriptHash` allowed in terms?
+
+An alternative design would allow UPLC terms to contain `ScriptHash`es

```

Comment:
We looked into something a bit like this in connection with Merklising PLC ASTs.  It was a non-starter because hashes are quite large (maybe 32 bytes) and theyr'e incompressible, so once you've got a few hashes in your script you've used up quite a lot of the size allowance.

Reply:
Ah... interesting. That strengthens the arguments against it. I will add a note to that effect.